Europython talk abstract about sciline. (#1)
scipp/presentations: europython2024/README.md

> # No more spaghetti code: build reusable data pipeline with Sciline

Reviewer: # Sciline: Assemble tasks graphs using type-hint-driven dependency injection

Author: I wanted a little bit of click-bait...

Reviewer: Think about the audience: task graphs, type hints, and dependency injection *is* clickbait ;)

## Abstract

> 

Reviewer: I think it is very important that we mention Dask prominently in the abstract.

Shall we just sit down together tomorrow to write the abstract?

Author: Yeah...! Let's do that tomorrow.

I was trying to focus on the 'way of working' that sciline can introduce between non-programmers and programmers.
So I didn't think about the part that only sw engineers will be interested...

> Have you dealt with scary spaghetti code that was not written by yourself?

Reviewer 2: I'm also not so sure `sciline` can help someone that has to make sense of someone else's old code?
It would if the old code was written with `sciline`, but it most probably wouldn't be?

Reviewer: We will work on the abstract together tomorrow, no more need to comment here ;)

Author: It's also good to collect opinions here in advance : D....

Or are you the one who wrote them?
Or do you want to have a visualization of the complicated data process but you do NOT want to use GUI-based frameworks?
If yes, [``sciline``](https://scipp.github.io/sciline) may be interesting to you.

> In this talk, we will introduce the ``sciline`` as a complex data process development framework. ``Sciline`` provides an automatic assembly of the task graph that can compute desired outputs or visualize the process to compute them. It is inspired by dependency injection frameworks, and it relies on Python’s type-hinting. As long as your interfaces are correctly type-hinted, you can use them with ``sciline`` right away!

Reviewer: Type-hinting is not sufficient, they must be unique, so this is misleading.

Author: Good point...

Reviewer 2: Yes, it sounds like we are claiming you can take any code and make it work with sciline, while in fact it requires making unique types for everything, which is (potentially) a lot of work.

> Sciline is developed for programmers and scientists to implement complicated data-processing workflows together in an efficient, consistent, clear, and sustainable way.

Author: ```suggestion
Sciline is developed for programmers and scientists to implement complicated data-processing workflows together in an efficient, clear and maintainable way.
```

# Sciline: Assemble task graphs using type-hint-driven dependency injection

## Abstract

Writing, testing, and maintaining complex data analysis workflows is hard.
Boiler-plate code may hide the actual analysis, making it hard to understand.
The code may be hard to test, leading to bugs.

Workflow management systems like [AiiDA](https://www.aiida.net/) or [Snakemake](https://snakemake.readthedocs.io/en/stable/)
can help with this by using a set of calculations or rules declaring their inputs and outputs.
They then automatically assemble a task graph, and run the tasks in the correct order to compute the desired outputs.

> Tasks may be complex with many internal computation steps but splitting them may not be feasible, e.g., due to overhead.

Reviewer 2: ```suggestion
Tasks may be complex, with many internal computation steps, but splitting them may not be feasible, e.g., due to overhead.
```

> [Sciline](https://scipp.github.io/sciline) can be a solution here.

Reviewer 2: I think it is not super clear here what sciline brings over the other packages. Right now, I think it sounds like it does the same as AiiDA and Snakemake.
Did you mean that sciline can help in cases where "splitting them may not be feasible"?

Reviewer: Yes, that is what we meant.

It is a pure Python package with no dependencies, and it does not require invasive changes to your existing Python code.

Sciline takes a *declarative* approach, inspired by [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) frameworks.
By relying on Python's type-hinting, a [domain-specific language](https://en.wikipedia.org/wiki/Domain-specific_language) is used to describe the workflow.
This enforces a clear expression of intent, aiding readability and enabling automatic assembly of a task graph that can compute desired outputs.

> The task graphs Sciline assembles can then be executed, e.g., using [Dask](https://dask.org/).

Reviewer 2: I think we could add more of the advantages that it brings, such as the ones I listed on slack:

- The ability to visualize the workflow as a graph (with lazy computation)
- The ability to inspect any intermediate result
- The absolute certainty that when you call compute on a pipeline, you are sure that the result you are getting is what you expect. It removes the annoying thing where in a long notebook, you have executed a cell a long way up, that modified your result, but you don't remember that you did that

Because a non-expert python user (like many scientists) would not know what "a declarative approach" is, what "dependency injection" is, or even what "type-hinting" is.

Some of these point are covered in the sentence

> This enforces a clear expression of intent, aiding readability and enabling automatic assembly of a task graph that can compute desired outputs.

but I wouldn't be surprised if many people don't really grasp what is really meant by it (e.g. "clear expression of intent" sounds quite generic).

Reviewer: > The absolute certainty that when you call compute on a pipeline, you are sure that the result you are getting is what you expect. It removes the annoying thing where in a long notebook, you have executed a cell a long way up, that modified your result, but you don't remember that you did that

I agree, but it seems too long for the abstract?

## Description

> This presentation will be in the form of a live demo of the Sciline. Sciline is an open-source project developed by the European Spallation Source under the BSD-3 licence. It is installable on Linux, Mac and Windows via pip and conda, and the documentation is hosted at https://scipp.github.io/sciline . The source code can be found at https://github.com/scipp . Co-authors: Jan-Lukas Wynen, Johannes Kasimir, Neil Vaytet, Simon Heybrock, Sunyoung Yoo.

Reviewer 2: ```suggestion
This presentation will be in the form of a live demo of Sciline. Sciline is an open-source project developed by the European Spallation Source under the BSD-3 licence. It is installable on Linux, Mac and Windows via pip and conda, and the documentation is hosted at https://scipp.github.io/sciline . The source code can be found at https://github.com/scipp/sciline . Co-authors: Jan-Lukas Wynen, Johannes Kasimir, Neil Vaytet, Simon Heybrock, Sunyoung Yoo.
```
